### PR TITLE
[Train] Don't use `NCCL_BLOCKING_WAIT`

### LINKED ISSUE #29419: [Core] NCCL_BLOCKING_WAIT=1 leads to performance regression
### What happened + What you expected to happen

Looks like NCCL_BLOCKING_WAIT is set to 1 for detecting node failure: https://github.com/ray-project/ray/blob/d99eff919bf785f911e4eebc87ddc4960344a139/python/ray/train/torch/config.py#L103 

But this will lead to significant performance regression, since CPU will stop scheduling new operator during the waiting. Can we just default it to zero to avoid the user setting in env variable?

More detail can be found in https://github.com/pytorch/pytorch/issues/87036

### Versions / Dependencies

2.0

### Reproduction script

NA

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by amogkam**
Signed-off-by: Amog Kamsetty <amogkamsetty@yahoo.com>

From the pytorch docs, we should use `NCCL_ASYNC_ERROR_HANDLING` instead.

Closes https://github.com/ray-project/ray/issues/29419

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch/config.py
Comment (jiaodong) [inline python/ray/train/torch/config.py:108]: can we update the test plan for failure behavior ? iiuc documentation says `NCCL_ASYNC_ERROR_HANDLING` is more performant but crashes the process, but `NCCL_BLOCKING_WAIT` will provide errors to the user which can be caught and handled --> this has implication of ray trainer's error handling semantics.

File: python/ray/train/torch/config.py
Comment (cadedaniel) [inline python/ray/train/torch/config.py:102]: Seems like we should respect `Only one of these two environment variables should be set.` from the docs you link, e.g. not set this if `NCCL_BLOCKING_WAIT` is set.

File: python/ray/train/torch/config.py
Comment (cadedaniel) [reply python/ray/train/torch/config.py:108]: +1, we should trigger this code path and make sure the crash output provides enough information to the user before merging. I don't think we can do much better than crashing unfortunately.

Comment (amogkam) [reply python/ray/train/torch/config.py:108]: Any suggestions on how to trigger this code path?

File: python/ray/train/torch/config.py
Comment (amogkam) [reply python/ray/train/torch/config.py:102]: good point! updated

File: python/ray/train/torch/config.py
Comment (cadedaniel) [reply python/ray/train/torch/config.py:112]: Launch data-parallel training (minimum two actors) that use NCCL to do the allreduce. Make one of the actors enter a while True: sleep loop so that it never enters the allreduce. Then, after 30 minutes, you'll see how PyTorch crashes the process. Will be even easier if you reduce the timeout ;)

Comment (amogkam) [reply python/ray/train/torch/config.py:112]: Yep, looks like an exception is being raised

```
(RayTrainWorker pid=13803) [E ProcessGroupNCCL.cpp:737] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=4, OpType=ALLREDUCE, Timeout(ms)=5000) ran for 7751 milliseconds before timing out.
(RayTrainWorker pid=13803) [E ProcessGroupNCCL.cpp:414] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data. To avoid this inconsistency, we are taking the entire process down.
(RayTrainWorker pid=13803) [2022-10-21 16:23:36,638 E 13803 13875] logging.cc:97: Unhandled exception: St13runtime_error. what(): [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=4, OpType=ALLREDUCE, Timeout(ms)=5000) ran for 7751 milliseconds before timing out.
(RayTrainWorker pid=13803) [2022-10-21 16:23:36,648 E 13803 13875] logging.cc:104: Stack trace: 
(RayTrainWorker pid=13803)  /home/ray/anaconda3/lib/python3.8/site-packages/ray/_raylet.so(+0xc74dda) [0x7f0934867dda] ray::operator<<()
(RayTrainWorker pid=13803) /home/ray/anaconda3/lib/python3.8/site-packages/ray/_raylet.so(+0xc77598) [0x7f093486a598] ray::TerminateHandler()
(RayTrainWorker pid=13803) /home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xacf6f) [0x7f0933b2af6f] __cxxabiv1::__terminate()
(RayTrainWorker pid=13803) /home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xacfb1) [0x7f0933b2afb1] __cxxabiv1::__unexpected()
(RayTrainWorker pid=13803) /home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xacf6c) [0x7f0933b2af6c] __cxxabiv1::__terminate()
(RayTrainWorker pid=13803) /home/ray/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so(_ZN4c10d16ProcessGroupNCCL8WorkNCCL15handleNCCLGuardEv+0x19f) [0x7efbc5ae2d4f] c10d::ProcessGroupNCCL::WorkNCCL::handleNCCLGuard()
(RayTrainWorker pid=13803) /home/ray/anaconda3/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so(_ZN4c10d16ProcessGroupNCCL15workCleanupLoopEv+0x199) [0x7efbc5ae71c9] c10d::ProcessGroupNCCL::workCleanupLoop()
(RayTrainWorker pid=13803) /home/ray/anaconda3/bin/../lib/libstdc++.so.6(+0xc9039) [0x7f0933b47039] execute_native_thread_routine
(RayTrainWorker pid=13803) /usr/lib/x86_64-linux-gnu/libpthread.so.0(+0x8609) [0x7f09354e5609] start_thread
(RayTrainWorker pid=13803) /usr/lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7f093540a133] __clone
(RayTrainWorker pid=13803) 
```

But the Ray Actor is still alive, causing training to hang. @rkooo567 do you know why the actor is not terminating when receiving this exception?

Comment (cadedaniel) [reply python/ray/train/torch/config.py:112]: Is the ray actor still alive? I think the process that contained the ray actor should be killed by `SIGABRT` https://github.com/ray-project/ray/blob/master/src/ray/util/logging.cc#L106

Comment (amogkam) [reply python/ray/train/torch/config.py:112]: yes the actor is still alive

**Comment by amogkam** [comment]
Blocked on https://github.com/ray-project/ray/issues/29576

File: python/ray/train/torch/config.py
Comment (amogkam) [reply python/ray/train/torch/config.py:111]: Added test...let's see if this passes

File: python/ray/train/tests/test_gpu.py
Comment (cadedaniel) [inline python/ray/train/tests/test_gpu.py:321]: nit: can we have a `time.sleep` here? that way we don't consume an entire CPU unnecessarily 🔥

**Review by cadedaniel — APPROVED** [review]

**Comment by jiaodong** [comment]
hmm wait what's the status of https://github.com/ray-project/ray/issues/29576 don't we need it first before changing our recovery ?

**Pull request merged**
